File: template_algos_cpp/tin_tout.cpp


int timer = 0;

void recur_fn(int node, int parent,  vector<vector<int>>& graph, vector<int>& arr, vector<int>& tin, vector<int>& tout, vector<int>& new_arr) {
    tin[node] = timer;
    timer += 1;
    new_arr.push_back(arr[node]);
    for (int child : graph[node]) {
        if (child == parent) continue;
        recur_fn(child,node,graph,arr,tin,tout,new_arr);
    }

    tout[node] = timer-1;
}

int solve(int N, vector<int>& arr, vector<vector<int>>& graph) {
    timer = 0;
    vector<int> tin(N);
    vector<int> tout(N);
    vector<int> new_arr;


    recur_fn(0,-1,graph,arr, tin,tout,new_arr);

    vector<int> pre(N+1);

    for (int i = 0; i < new_arr.size(); ++i) {
        pre[i+1] = max(pre[i],new_arr[i]);
    }
    vector<int> suff(N+1);
    for (int i = new_arr.size()-1; i >= 0; i--) {
        suff[i] = max(suff[i+1],new_arr[i]);
    }

    int ans = 0;
    int max_weight = 0;
    for (int i = 0; i < arr.size(); ++i) {
        if (max(pre[tin[i]],suff[tout[i]+1]) > arr[i]) {
            if (arr[i] > max_weight) {
                max_weight = arr[i];
                ans = i+1;
            }
        }
    }

    return ans;

}

File: template_algos/tin_tout/tin_tout.py

def solve(N ,arr ,graph):

    timer = 0
    new_arr = []
    tin = [0 ] *N
    tout = [0 ] *N

    stack = [(0, -1, False)]

    while stack:
        node, par, visited = stack.pop()

        if not visited:
            tin[node] = timer
            timer += 1
            new_arr.append(arr[node])

            stack.append((node, par, True))

            for child in reversed(graph[node]):
                if child != par:
                    stack.append((child, node, False))
        else:
            tout[node] = timer - 1

    pre = [0 ] *( N +1)
    suf = [0 ] *( N +1)

    for i in range(len(new_arr)):
        pre[ i +1] = max(pre[i] ,new_arr[i])
    for i in range(len(new_arr ) -1 ,-1 ,-1):
        suf[i] = max(suf[ i +1] ,new_arr[i])


    max_weight = 0
    ans = 0

    for i in range(len(arr)):
        if max(pre[tin[i]] ,suf[tout[i ] +1]) > arr[i]:
            if arr[i] > max_weight:
                ans = i+ 1
                max_weight = arr[i]
    return ans